File: 0456-Find-132-Pattern/0456.java
class Solution {
  public boolean find132pattern(int[] nums) {
    int n = nums.length;
    Deque<Integer> candidateK = new LinkedList<Integer>();
    candidateK.push(nums[n - 1]);
    int maxK = Integer.MIN_VALUE;

    for (int i = n - 2; i >= 0; --i) {
      if (nums[i] < max) {
        return true;
      }
      while (!candidateK.isEmpty() && nums[i] > candidateK.peek()) {
        maxK = candidateK.pop();
      }
      if (nums[i] > maxK) {
        candidateK.push(nums[i]);
      }
    }
    return false;
  }
}

File: 0456-Find-132-Pattern/0456.py
class Solution:
  def find132pattern(self, nums: List[int]) -> bool:
    n = len(nums)
    candidate_k = [nums[n - 1]]
    max_k = float('-inf')

    for i in range(n - 2, -1, -1):
      if nums[i] < max_k:
        return True
      while candidate_k and num[i] > candidate_k[-1]:
        max_k = candidate_k[-1]
        candidate_k.pop()
      if nums[i] > max_k:
        candidate_k.append(nums[i])

    return False